### PR TITLE
Reworked crate structure isolating core and rebooted hybrid crypto

### PR DESCRIPTION
**Pull request by bgrieder**
- reworked crate structure isolating core and interfaces (Cosmian specific traits + FFI)
 - rebooted hybrid crypto exposing a FFI for symmetric encryption with AES 156 GCM

### CODE REVIEW

File: src/interfaces/hybrid_crypto/structs.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
+use std::marker::PhantomData;
```
Comment (Manuthor) [inline src/interfaces/hybrid_crypto/structs.rs:1]: do we keep this file? (apparently not used anymore)

File: src/interfaces/asymmetric_crypto.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S> AbeCrypto<S>
+where
+    S: AbeScheme + std::marker::Sync + std::marker::Send,
+{
+    // fn ensure_engine(&self) -> anyhow::Result<&Engine<S>> {
```
Comment (Manuthor) [inline src/interfaces/asymmetric_crypto.rs:97]: probably remove this

File: src/interfaces/ffi/structs.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
+// ---------------------------------------
```
Comment (Manuthor) [inline src/interfaces/ffi/structs.rs:1]: this file does not seem to be used anymore

File: src/interfaces/old/master_key.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
+use std::{fmt::Display, vec::Vec};
```
Comment (Manuthor) [inline src/interfaces/old/master_key.rs:1]: do we keep the `old` folder?

File: src/interfaces/asymmetric_crypto.rs [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    //     self.engine.as_ref().ok_or_else(|| {
+    //         anyhow::anyhow!(
+    //             "The Policy must first be set on ABE scheme using set_scheme_parameters"
+    //         )
+    //     })
+    // }
+
+    /// Generate an ABE master key pair for the Policy, returning a triple
+    /// (private key, public key, public delegation key)
+    fn generate_master_keys(&self, policy: &Policy) -> anyhow::Result<AbeMasterKeys<S>> {
```
Comment (ThibsG) [inline src/interfaces/asymmetric_crypto.rs:107]:
```suggestion
    [cfg(test)]
    fn generate_master_keys(&self, policy: &Policy) -> anyhow::Result<AbeMasterKeys<S>> {
```

File: src/interfaces/ffi/README.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Files under this module allow exposing a C ABI interface for integration to Python, Java, C, C7+ etc...
```
Comment (ThibsG) [inline src/interfaces/ffi/README.md:1]: typo?
```suggestion
Files under this module allow exposing a C ABI interface for integration to Python, Java, C, C++ etc...
```

File: src/interfaces/ffi/error.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    // Build a CString that will cleanup NULL bytes in the middle if needed
+    let cs = ffi_unwrap!(
+        CString::new(err.map_or("".to_string(), |e| e.to_string())),
+        "failed to convert error to CString"
+    );
+    // the CString as bytes
+    let bytes = cs.as_bytes();
+
+    // leave a space for a null byte at the end if the string exceeds the buffer
+    // The actual bytes size, not taking into accourt the final NULL
```
Comment (ThibsG) [inline src/interfaces/ffi/error.rs:198]:
```suggestion
    // The actual bytes size, not taking into account the final NULL
```

File: src/interfaces/hybrid_crypto/tests/mod.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+use cosmian_crypto_base::{
+    hybrid_crypto::Metadata,
+    symmetric_crypto::{aes_256_gcm_pure::Aes256GcmCrypto, Key},
+};
+use serde_json::Value;
+
+type PublicKey = <Gpsw<Bls12_381> as AbeScheme>::MasterPublicKey;
+type UserDecryptionKey = <Gpsw<Bls12_381> as AbeScheme>::UserDecryptionKey;
+
+// maximum clear text size that can be safely encrypted with AES GCM (using a a single random nonce)
```
Comment (ThibsG) [inline src/interfaces/hybrid_crypto/tests/mod.rs:21]:
```suggestion
// maximum clear text size that can be safely encrypted with AES GCM (using a single random nonce)
```

File: src/core/tests/engine_tests.rs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    policy::{attr, Policy},
+    Engine,
+};
+
+type PublicKey = <Gpsw<Bls12_381> as AbeScheme>::MasterPublicKey;
+type UserDecryptionKey = <Gpsw<Bls12_381> as AbeScheme>::UserDecryptionKey;
+
+#[test]
+pub fn symmetric_key_test() {
+    //-> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/core/tests/engine_tests.rs:13]: Done

File: src/core/engine.rs [hunk 10]
```diff
@@ -0,0 +1,153 @@
+#![allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/core/engine.rs:1]: Changed the visibility - and removed the annotation

File: src/core/engine.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        priv_key: &S::MasterPrivateKey,
+        access_policy: &AccessPolicy,
+    ) -> Result<S::UserDecryptionKey, FormatErr> {
+        let msp = policy.to_msp(access_policy)?;
+        self.sch.key_generation(&msp, priv_key)
+    }
+
+    /// Allows a user to generate a new key for a more restrictive policy
+    ///
+    /// A more restrictive policy is a policy for which when it is satisfy, the less
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/core/engine.rs:62]: Tried a bette explanation

File: src/core/engine.rs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    /// restrictive also. In other words, we can only modify a policy by changing
+    /// an `Or` node by either an `And` or replace it by one of its child.
+    ///
+    /// Remark: It is also possible to merge 2 keys by `Or` node, this latter
+    /// functionality is not yet supported
+    // Not exposed to public FFI API yet: hence he dead_code
+    #[allow(dead_code)]
+    pub fn delegate_user_key(
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/core/engine.rs:70]: I changed the module visibility. This should be exposed

File: src/interfaces/ffi/structs.rs (hunk 3 above)
Comment (bgrieder) [reply src/interfaces/ffi/structs.rs:1]: Yes. Right. That was my attempt at using opaque objects through the FFI but it fails. Removing it

File: src/interfaces/hybrid_crypto/structs.rs (hunk 1 above)
Comment (bgrieder) [reply src/interfaces/hybrid_crypto/structs.rs:1]: No rigt. Same comment as before. Failed attempt at using opaque objects on the FFI. Removing it

File: src/interfaces/old/master_key.rs (hunk 4 above)
Comment (bgrieder) [reply src/interfaces/old/master_key.rs:1]: NO. Removed

File: src/interfaces/asymmetric_crypto.rs (hunk 5 above)
Comment (bgrieder) [reply src/interfaces/asymmetric_crypto.rs:107]: Not sure whether this should or should not be publibly available. Will limit visibility to test as suggested.

File: Cargo.toml [hunk 13]
```diff
@@ -15,21 +15,28 @@ version = "0.2.4"
 
 [lib]
 name = "abe_gpsw"
+crate-type = ["lib", "cdylib"]
+# The cdylib is only intersting if the `--features ffi` flag is set on build
+# This does not seem to be actionnable conditionnally https://github.com/rust-lang/cargo/issues/4881
```
Comment (ThibsG) [inline Cargo.toml:20]:
```suggestion
# The cdylib is only interesting if the `--features ffi` flag is set on build
# This does not seem to be actionable conditionally https://github.com/rust-lang/cargo/issues/4881
```

File: src/core/engine.rs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            S::MasterPublicKey,
+            S::MasterPublicDelegationKey,
+        ),
+        FormatErr,
+    > {
+        self.sch.generate_master_key(policy.max_attr())
+    }
+
+    /// Generate a user decryption key
+    /// from the supplied Master Private Key andAccess Policy
```
Comment (ThibsG) [inline src/core/engine.rs:50]:
```suggestion
    /// from the supplied Master Private Key and Access Policy
```

File: src/core/engine.rs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+        access_policy: &AccessPolicy,
+    ) -> Result<S::UserDecryptionKey, FormatErr> {
+        let msp = policy.to_msp(access_policy)?;
+        self.sch.key_generation(&msp, priv_key)
+    }
+
+    /// Allows a user to generate a new key for a more restrictive policy
+    ///
+    /// A more restrictive policy is a policy that must always satisfy
+    /// the original policy when satisfied.. In other words, we can only modify a policy by changing
```
Comment (ThibsG) [inline src/core/engine.rs:64]:
```suggestion
    /// the original policy when satisfied. In other words, we can only modify a policy by changing
```

File: src/core/engine.rs [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    /// an `Or` node by either an `And` or replace it by one of its child.
+    ///
+    /// Remark: It is also possible to merge 2 keys by `Or` node, this latter
+    /// functionality is not yet supported
+    // Not exposed to public FFI API yet: hence he dead_code
+    #[allow(dead_code)]
```
Comment (ThibsG) [inline src/core/engine.rs:70]: This is now useless
```suggestion
```

**Pull request merged**